File: sdks/ruby/lib/condux/modules.rb
# frozen_string_literal: true

module Condux
  # The runtime dependency inventory that rides events (ADR-0041): which gem versions are actually
  # loaded, as opposed to which ones a Gemfile declares. The relay indexes this so a security advisory
  # can be answered with "and you are running 2.3.0 in production" rather than only "your lockfile
  # says so".
  #
  # <b>This one is collected lazily, unlike the other SDKs, and that is not an inconsistency.</b>
  # Gem.loaded_specs reports the gems that have been ACTIVATED, not the ones installed. Measured in a
  # bare interpreter it holds a single entry, because nothing has been required yet. Collecting during
  # Condux.init, which an application typically calls early in boot, would therefore report almost
  # nothing. Collecting at capture time reports what was genuinely loaded by the moment the error
  # happened, which is both correct and the truest reading of what this feature claims to measure. It
  # costs nothing: loaded_specs is an in-memory hash, with no filesystem behind it.
  module Modules
    # The most entries carried on one event. The cap applies after sorting, so which entries survive is
    # stable across events rather than varying with hash order: the server sees one consistent set
    # instead of a shifting sample.
    MAX_MODULES = 1000

    # How long to wait before repeating the inventory on another event.
    #
    # This is what makes the feature affordable. The server deduplicates a release's inventory down to
    # one row per package per day, so attaching the whole map to every event would spend bytes for
    # nothing. Repeating on an interval rather than sending once keeps the robustness that every-event
    # buys: the event carrying the inventory can be dropped by a rate limit or a quota rejection
    # before anything parses it, so a single attempt per process would lose that day's inventory.
    MODULES_INTERVAL_SECONDS = 15 * 60

    module_function

    # The loaded gems as a name to version hash, empty when they cannot be read.
    #
    # Never raises. This runs inside capture, which is already handling somebody's error, so a broken
    # spec costs that entry and nothing else.
    def collect
      found = {}
      Gem.loaded_specs.each do |name, spec|
        version = spec.version.to_s
        found[name.to_s] = version if name && !name.to_s.empty? && !version.empty?
      rescue StandardError
        # One unreadable spec is not a reason to report nothing about the rest.
        next
      end
      found
    rescue StandardError, NameError
      # No RubyGems at all (a packaged binary, a trimmed runtime). Unknown is the honest answer.
      {}
    end

    # The inventory's contribution to an event: the full map on the first event and then at most once
    # per MODULES_INTERVAL_SECONDS, and an empty hash otherwise, so an event that carries nothing keeps
    # its exact previous wire shape.
    def fields(now)
      return {} if @last_attached_at && (now - @last_attached_at) < MODULES_INTERVAL_SECONDS

      loaded = collect
      return {} if loaded.empty?

      @last_attached_at = now
      { "modules" => loaded.sort.first(MAX_MODULES).to_h }
    end

    # Reset the interval. Tests only; a process loads its gems once.
    def reset
      @last_attached_at = nil
    end
  end
end

File: sdks/python/src/condux/modules.py
"""The runtime dependency inventory that rides events (ADR-0041).

Which package versions are actually installed alongside the running process, as opposed to which ones
a requirements file declares. The relay indexes this so a security advisory can be answered with "and
you are running 2.19.0 in production" rather than only "your lockfile says so".

Reading the installed distributions is what answers that: a requirements file states ranges, and what
matters is what resolution actually put in the environment. ``importlib.metadata`` is the stdlib's own
view of exactly that, so there is nothing to parse and nothing to guess.
"""

import time
from typing import Dict, Mapping, Optional

# The most entries carried on one event. A large environment runs past this, and the cap applies after
# sorting, so which entries survive is stable across events rather than varying with iteration order:
# the server sees one consistent set instead of a shifting sample.
MAX_MODULES = 1000

# How long to wait before repeating the inventory on another event.
#
# This is what makes the feature affordable. The map is repeated on every event that carries it, and
# the server deduplicates a release's inventory down to one row per package per day, so attaching it
# to every event would spend bytes for nothing. Repeating on an interval rather than sending once
# keeps the robustness that every-event buys: the event carrying the inventory can be dropped by a
# rate limit or a quota rejection before anything parses it, so a single attempt per process would
# lose that day's inventory outright.
MODULES_INTERVAL_SECONDS = 15 * 60

_modules: Optional[Dict[str, str]] = None
_last_attached_at: Optional[float] = None


def collect_modules() -> Dict[str, str]:
    """The installed distributions as a name to version mapping, empty when it cannot be read.

    Never raises. This runs during ``init`` in an application that never asked for it, so a broken
    distribution or an unreadable metadata directory costs that entry, or at worst the inventory, and
    never the caller's startup.
    """
    found: Dict[str, str] = {}
    try:
        from importlib.metadata import distributions

        for dist in distributions():
            try:
                name = dist.metadata["Name"]
                version = dist.version
            except Exception:
                # One unreadable distribution is not a reason to report nothing about the rest.
                continue
            if name and version:
                # setdefault, so the first copy on the path wins, which is the one an import resolves.
                found.setdefault(str(name), str(version))
    except Exception:
        return {}
    return found


def set_modules(modules: Optional[Mapping[str, str]]) -> None:
    """Declare the installed package versions for subsequent events; ``None`` clears them."""
    global _modules, _last_attached_at

    # A fresh declaration is news, so let the next event carry it rather than waiting out an interval
    # started by the previous inventory.
    _last_attached_at = None

    if modules is None:
        _modules = None
        return

    entries = sorted(
        (str(name), str(version))
        for name, version in modules.items()
        if name and version
    )[:MAX_MODULES]
    _modules = dict(entries) if entries else None


def modules_field(now: Optional[float] = None) -> Dict[str, Dict[str, str]]:
    """The inventory's contribution to an event.

    The full map on the first event and then at most once per :data:`MODULES_INTERVAL_SECONDS`, and an
    empty dict otherwise, so an event that carries nothing keeps its exact previous wire shape.
    """
    global _last_attached_at

    if _modules is None:
        return {}

    moment = time.time() if now is None else now
    if _last_attached_at is not None and moment - _last_attached_at < MODULES_INTERVAL_SECONDS:
        return {}

    _last_attached_at = moment
    return {"modules": dict(_modules)}


def clear_modules() -> None:
    """Reset the inventory and its interval. Tests only; a process has one environment."""
    global _modules, _last_attached_at
    _modules = None
    _last_attached_at = None
